File: src/main/java/org/trostheide/generativeArt/ReactionDiffusionGenerator.java
package org.trostheide.generativeArt;

import org.trostheide.generativeArt.core.ArtGenerator;
import org.trostheide.generativeArt.core.ParameterDefinition;

import java.util.List;
import java.util.Map;

public class ReactionDiffusionGenerator implements ArtGenerator {

    private int width = 200;
    private int height = 200;
    private int scale = 2;
    private int iterations = 8000;
    private double da = 1.0;
    private double db = 0.5;
    private double f = 0.055;
    private double k = 0.062;

    @Override
    public String getId() {
        return "reaction-diffusion";
    }

    @Override
    public String getDisplayName() {
        return "Reaction Diffusion (Gray-Scott)";
    }

    public List<ParameterDefinition> getParameterDefinitions() {
        return List.of(
                ParameterDefinition.selection("Preset", "Custom", 
                    java.util.Arrays.asList("Custom", "Coral Growth", "Mitosis", "Mazes", "Moving Spots"), 
                    "Select a predefined style"),
                ParameterDefinition.doubleVal("Feed Rate", 0.055, 0.01, 0.1, "Feed rate (f)"),
                ParameterDefinition.doubleVal("Kill Rate", 0.062, 0.01, 0.1, "Kill rate (k)"),
                ParameterDefinition.integer("Iterations", 8000, 1000, 20000, "Simulation steps"),
                ParameterDefinition.doubleVal("Threshold", 0.25, 0.1, 0.9, "Start iso-contour threshold"),
                ParameterDefinition.integer("Scale", 2, 1, 5, "Output scale"),
                ParameterDefinition.integer("Colors", 1, 1, 6, "Number of layers (contours)"));
    }

    @Override
    public boolean onParameterChanged(String paramName, Object newValue, Map<String, Object> currentValues) {
        if ("Preset".equals(paramName) && newValue instanceof String) {
            String preset = (String) newValue;
            switch (preset) {
                case "Coral Growth":
                    currentValues.put("Feed Rate", 0.0545);
                    currentValues.put("Kill Rate", 0.062);
                    currentValues.put("Iterations", 10000);
                    currentValues.put("Threshold", 0.3);
                    currentValues.put("Scale", 2);
                    currentValues.put("Colors", 2);
                    return true;
                case "Mitosis":
                    currentValues.put("Feed Rate", 0.0367);
                    currentValues.put("Kill Rate", 0.0649);
                    currentValues.put("Iterations", 8000);
                    currentValues.put("Threshold", 0.25);
                    currentValues.put("Scale", 2);
                    currentValues.put("Colors", 1);
                    return true;
                case "Mazes":
                    currentValues.put("Feed Rate", 0.029);
                    currentValues.put("Kill Rate", 0.057);
                    currentValues.put("Iterations", 10000);
                    currentValues.put("Threshold", 0.2);
                    currentValues.put("Scale", 2);
                    currentValues.put("Colors", 1);
                    return true;
                case "Moving Spots":
                    currentValues.put("Feed Rate", 0.014);
                    currentValues.put("Kill Rate", 0.054);
                    currentValues.put("Iterations", 12000);
                    currentValues.put("Threshold", 0.4);
                    currentValues.put("Scale", 3);
                    currentValues.put("Colors", 3);
                    return true;
                case "Custom":
                default:
                    return false;
            }
        }
        if (!"Preset".equals(paramName) && !"Custom".equals(currentValues.get("Preset"))) {
            currentValues.put("Preset", "Custom");
            return true;
        }
        return false;
    }

    @Override
    public String generate(Map<String, Object> params) {

        int totalW = width * scale;
        int totalH = height * scale;
        double startThreshold = (double) params.getOrDefault("Threshold", 0.25);
        int numColors = (int) params.getOrDefault("Colors", 1);

        // Ensure simulation is run
        Simulation sim = new Simulation(width, height, da, db, f, k);
        sim.run(iterations);

        org.trostheide.generativeArt.core.SvgCanvas canvas = new org.trostheide.generativeArt.core.SvgCanvas(totalW,
                totalH, numColors);

        // Generate contours for each layer
        for (int i = 0; i < numColors; i++) {
            // Vary threshold slightly for each layer to create concentric rings
            // e.g., 0.2, 0.25, 0.3...
            double threshold = startThreshold + (i * 0.05);
            if (threshold >= 1.0)
                threshold = 0.99;

            String paths = MarchingSquares.vectorize(sim.b, width, height, threshold);

            String groupContent = String.format(java.util.Locale.US,
                    "<g transform='scale(%d)'>\n" +
                            "<path d='%s' fill='none' stroke-width='1.5' stroke-linecap='round' />\n" +
                            "</g>",
                    scale, paths);

            canvas.addRaw(i, groupContent);
        }

        return canvas.toSvg();
    }

    // --- Simulation ---
    private static class Simulation {
        int w, h;
        double f, k, da, db;
        double[] a, b, nextA, nextB;
        private static final double DT = 1.0;

        public Simulation(int w, int h, double da, double db, double f, double k) {
            this.w = w;
            this.h = h;
            this.da = da;
            this.db = db;
            this.f = f;
            this.k = k;

            int size = w * h;
            a = new double[size];
            b = new double[size];
            nextA = new double[size];
            nextB = new double[size];

            // Init
            for (int i = 0; i < size; i++) {
                a[i] = 1.0;
                b[i] = 0.0;
            }

            // Seed
            int seedRadius = 15;
            for (int i = w / 2 - seedRadius; i < w / 2 + seedRadius; i++) {
                for (int j = h / 2 - seedRadius; j < h / 2 + seedRadius; j++) {
                    b[i + j * w] = 1.0;
                }
            }
        }

        public void run(int steps) {
            double center = -1.0;
            double adj = 0.2;
            double diag = 0.05;

            for (int t = 0; t < steps; t++) {
                for (int x = 1; x < w - 1; x++) {
                    for (int y = 1; y < h - 1; y++) {
                        int i = x + y * w;
                        double u = a[i];
                        double v = b[i];

                        double lu = (a[i - 1] * adj) + (a[i + 1] * adj) + (a[i - w] * adj) + (a[i + w] * adj) +
                                (a[i - w - 1] * diag) + (a[i - w + 1] * diag) + (a[i + w - 1] * diag)
                                + (a[i + w + 1] * diag) +
                                (u * center);
                        double lv = (b[i - 1] * adj) + (b[i + 1] * adj) + (b[i - w] * adj) + (b[i + w] * adj) +
                                (b[i - w - 1] * diag) + (b[i - w + 1] * diag) + (b[i + w - 1] * diag)
                                + (b[i + w + 1] * diag) +
                                (v * center);

                        double reaction = u * v * v;

                        nextA[i] = u + (da * lu - reaction + f * (1 - u)) * DT;
                        nextB[i] = v + (db * lv + reaction - (k + f) * v) * DT;

                        if (nextA[i] < 0)
                            nextA[i] = 0;
                        else if (nextA[i] > 1)
                            nextA[i] = 1;
                        if (nextB[i] < 0)
                            nextB[i] = 0;
                        else if (nextB[i] > 1)
                            nextB[i] = 1;
                    }
                }
                double[] temp = a;
                a = nextA;
                nextA = temp;
                temp = b;
                b = nextB;
                nextB = temp;
            }
        }
    }

    // --- Marching Squares ---
    private static class MarchingSquares {
        public static String vectorize(double[] data, int w, int h, double threshold) {
            StringBuilder path = new StringBuilder();

            for (int y = 0; y < h - 1; y++) {
                for (int x = 0; x < w - 1; x++) {
                    int i = x + y * w;
                    int state = 0;
                    if (data[i] > threshold)
                        state |= 8;
                    if (data[i + 1] > threshold)
                        state |= 4;
                    if (data[i + w + 1] > threshold)
                        state |= 2;
                    if (data[i + w] > threshold)
                        state |= 1;

                    if (state == 0 || state == 15)
                        continue;

                    double topX = x + 0.5, topY = y;
                    double rightX = x + 1, rightY = y + 0.5;
                    double botX = x + 0.5, botY = y + 1;
                    double leftX = x, leftY = y + 0.5;

                    switch (state) {
                        case 1:
                            path.append(line(leftX, leftY, botX, botY));
                            break;
                        case 2:
                            path.append(line(botX, botY, rightX, rightY));
                            break;
                        case 3:
                            path.append(line(leftX, leftY, rightX, rightY));
                            break;
                        case 4:
                            path.append(line(topX, topY, rightX, rightY));
                            break;
                        case 5:
                            path.append(line(leftX, leftY, topX, topY));
                            path.append(line(botX, botY, rightX, rightY));
                            break;
                        case 6:
                            path.append(line(topX, topY, botX, botY));
                            break;
                        case 7:
                            path.append(line(leftX, leftY, topX, topY));
                            break;
                        case 8:
                            path.append(line(leftX, leftY, topX, topY));
                            break;
                        case 9:
                            path.append(line(topX, topY, botX, botY));
                            break;
                        case 10:
                            path.append(line(topX, topY, rightX, rightY));
                            path.append(line(botX, botY, leftX, leftY));
                            break;
                        case 11:
                            path.append(line(topX, topY, rightX, rightY));
                            break;
                        case 12:
                            path.append(line(leftX, leftY, rightX, rightY));
                            break;
                        case 13:
                            path.append(line(botX, botY, rightX, rightY));
                            break;
                        case 14:
                            path.append(line(leftX, leftY, botX, botY));
                            break;
                    }
                }
            }
            return path.toString();
        }

        private static String line(double x1, double y1, double x2, double y2) {
            return "M " + ((float) x1) + " " + ((float) y1) + " L " + ((float) x2) + " " + ((float) y2) + " ";
        }
    }
}

File: web/src/generators/ReactionDiffusionGenerator.js
import { Generator } from '../core/Generator';
import { ParameterDefinition } from '../core/ParameterDefinition';
import { SvgCanvas } from '../core/SvgCanvas';

export class ReactionDiffusionGenerator extends Generator {
    constructor() {
        super();
        this.width = 200;
        this.height = 200;
        this.scale = 2;
        this.iterations = 2000;
        this.da = 1.0;
        this.db = 0.5;
        this.f = 0.055;
        this.k = 0.062;
    }

    getId() {
        return "reaction-diffusion";
    }

    getDisplayName() {
        return "Reaction Diffusion (Gray-Scott)";
    }

    getParameterDefinitions() {
        return [
            ParameterDefinition.selection("Preset", "Custom", ["Custom", "Coral Growth", "Mitosis", "Mazes", "Moving Spots"], "Select a predefined style"),
            ParameterDefinition.doubleVal("Feed Rate", 0.055, 0.01, 0.1, "Feed rate (f)"),
            ParameterDefinition.doubleVal("Kill Rate", 0.062, 0.01, 0.1, "Kill rate (k)"),
            ParameterDefinition.integer("Iterations", 2000, 1000, 20000, "Simulation steps; higher values are slower and best triggered with Generate"),
            ParameterDefinition.doubleVal("Threshold", 0.25, 0.1, 0.9, "Start iso-contour threshold"),
            ParameterDefinition.integer("Scale", 2, 1, 5, "Output scale"),
            ParameterDefinition.integer("Colors", 1, 1, 6, "Number of layers (contours)")
        ];
    }

    onParameterChanged(paramName, newValue, currentValues) {
        if (paramName === "Preset" && typeof newValue === "string") {
            const preset = newValue;
            switch (preset) {
                case "Coral Growth":
                    currentValues["Feed Rate"] = 0.0545;
                    currentValues["Kill Rate"] = 0.062;
                    currentValues["Iterations"] = 10000;
                    currentValues["Threshold"] = 0.3;
                    currentValues["Scale"] = 2;
                    currentValues["Colors"] = 2;
                    return true;
                case "Mitosis":
                    currentValues["Feed Rate"] = 0.0367;
                    currentValues["Kill Rate"] = 0.0649;
                    currentValues["Iterations"] = 8000;
                    currentValues["Threshold"] = 0.25;
                    currentValues["Scale"] = 2;
                    currentValues["Colors"] = 1;
                    return true;
                case "Mazes":
                    currentValues["Feed Rate"] = 0.029;
                    currentValues["Kill Rate"] = 0.057;
                    currentValues["Iterations"] = 10000;
                    currentValues["Threshold"] = 0.2;
                    currentValues["Scale"] = 2;
                    currentValues["Colors"] = 1;
                    return true;
                case "Moving Spots":
                    currentValues["Feed Rate"] = 0.014;
                    currentValues["Kill Rate"] = 0.054;
                    currentValues["Iterations"] = 12000;
                    currentValues["Threshold"] = 0.4;
                    currentValues["Scale"] = 3;
                    currentValues["Colors"] = 3;
                    return true;
                case "Custom":
                default:
                    return false;
            }
        }
        if (paramName !== "Preset" && currentValues["Preset"] !== "Custom") {
            currentValues["Preset"] = "Custom";
            return true;
        }
        return false;
    }

    generate(params) {
        // Read Params
        this.f = params["Feed Rate"] || 0.055;
        this.k = params["Kill Rate"] || 0.062;
        this.iterations = params["Iterations"] || 8000;
        const startThreshold = params["Threshold"] || 0.25;
        this.scale = params["Scale"] || 2;
        const numColors = params["Colors"] || 1;

        const totalW = this.width * this.scale;
        const totalH = this.height * this.scale;

        // Run Simulation
        const sim = new Simulation(this.width, this.height, this.da, this.db, this.f, this.k);
        sim.run(this.iterations);

        const canvas = new SvgCanvas(totalW, totalH, numColors);

        // Generate contours
        for (let i = 0; i < numColors; i++) {
            let threshold = startThreshold + (i * 0.05);
            if (threshold >= 1.0) threshold = 0.99;

            const paths = MarchingSquares.vectorize(sim.b, this.width, this.height, threshold);

            // Group transform logic
            const groupContent = `<g transform='scale(${this.scale})'>\n` +
                `<path d='${paths}' fill='none' stroke-width='1.5' stroke-linecap='round' />\n` +
                `</g>`;

            canvas.addRaw(i, groupContent);
        }

        return canvas.toSvg();
    }
}

class Simulation {
    constructor(w, h, da, db, f, k) {
        this.w = w;
        this.h = h;
        this.da = da;
        this.db = db;
        this.f = f;
        this.k = k;

        const size = w * h;
        this.a = new Float64Array(size);
        this.b = new Float64Array(size);
        this.nextA = new Float64Array(size);
        this.nextB = new Float64Array(size);

        // Init
        for (let i = 0; i < size; i++) {
            this.a[i] = 1.0;
            this.b[i] = 0.0;
        }

        // Seed
        const seedRadius = 15;
        const centerX = Math.floor(w / 2);
        const centerY = Math.floor(h / 2);
        for (let i = centerX - seedRadius; i < centerX + seedRadius; i++) {
            for (let j = centerY - seedRadius; j < centerY + seedRadius; j++) {
                this.b[i + j * w] = 1.0;
            }
        }
    }

    run(steps) {
        const center = -1.0;
        const adj = 0.2;
        const diag = 0.05;
        const DT = 1.0;

        for (let t = 0; t < steps; t++) {
            for (let x = 1; x < this.w - 1; x++) {
                for (let y = 1; y < this.h - 1; y++) {
                    const i = x + y * this.w;
                    const u = this.a[i];
                    const v = this.b[i];

                    const lu = (this.a[i - 1] * adj) + (this.a[i + 1] * adj) + (this.a[i - this.w] * adj) + (this.a[i + this.w] * adj) +
                        (this.a[i - this.w - 1] * diag) + (this.a[i - this.w + 1] * diag) + (this.a[i + this.w - 1] * diag) + (this.a[i + this.w + 1] * diag) +
                        (u * center);

                    const lv = (this.b[i - 1] * adj) + (this.b[i + 1] * adj) + (this.b[i - this.w] * adj) + (this.b[i + this.w] * adj) +
                        (this.b[i - this.w - 1] * diag) + (this.b[i - this.w + 1] * diag) + (this.b[i + this.w - 1] * diag) + (this.b[i + this.w + 1] * diag) +
                        (v * center);

                    const reaction = u * v * v;

                    let nextValA = u + (this.da * lu - reaction + this.f * (1 - u)) * DT;
                    let nextValB = v + (this.db * lv + reaction - (this.k + this.f) * v) * DT;

                    // Clamp
                    if (nextValA < 0) nextValA = 0;
                    else if (nextValA > 1) nextValA = 1;

                    if (nextValB < 0) nextValB = 0;
                    else if (nextValB > 1) nextValB = 1;

                    this.nextA[i] = nextValA;
                    this.nextB[i] = nextValB;
                }
            }

            // Swap
            let temp = this.a;
            this.a = this.nextA;
            this.nextA = temp;

            temp = this.b;
            this.b = this.nextB;
            this.nextB = temp;
        }
    }
}

class MarchingSquares {
    static vectorize(data, w, h, threshold) {
        let path = "";

        for (let y = 0; y < h - 1; y++) {
            for (let x = 0; x < w - 1; x++) {
                const i = x + y * w;
                let state = 0;

                if (data[i] > threshold) state |= 8;
                if (data[i + 1] > threshold) state |= 4;
                if (data[i + w + 1] > threshold) state |= 2;
                if (data[i + w] > threshold) state |= 1;

                if (state === 0 || state === 15) continue;

                const topX = x + 0.5, topY = y;
                const rightX = x + 1, rightY = y + 0.5;
                const botX = x + 0.5, botY = y + 1;
                const leftX = x, leftY = y + 0.5;

                switch (state) {
                    case 1: path += this.line(leftX, leftY, botX, botY); break;
                    case 2: path += this.line(botX, botY, rightX, rightY); break;
                    case 3: path += this.line(leftX, leftY, rightX, rightY); break;
                    case 4: path += this.line(topX, topY, rightX, rightY); break;
                    case 5:
                        path += this.line(leftX, leftY, topX, topY);
                        path += this.line(botX, botY, rightX, rightY);
                        break;
                    case 6: path += this.line(topX, topY, botX, botY); break;
                    case 7: path += this.line(leftX, leftY, topX, topY); break;
                    case 8: path += this.line(leftX, leftY, topX, topY); break;
                    case 9: path += this.line(topX, topY, botX, botY); break;
                    case 10:
                        path += this.line(topX, topY, rightX, rightY);
                        path += this.line(botX, botY, leftX, leftY);
                        break;
                    case 11: path += this.line(topX, topY, rightX, rightY); break;
                    case 12: path += this.line(leftX, leftY, rightX, rightY); break;
                    case 13: path += this.line(botX, botY, rightX, rightY); break;
                    case 14: path += this.line(leftX, leftY, botX, botY); break;
                }
            }
        }
        return path;
    }

    static line(x1, y1, x2, y2) {
        return `M ${x1} ${y1} L ${x2} ${y2} `;
    }
}
